48-kori-bustard/technical_specification.md

```diff
@@ -0,0 +1,171 @@
+- Initial implementation of NOS
+- Notification Service Idempotence
+- Addition of new event schemas to ghga-event-schemas
+- Replace ARS notification events
+
+
+## Notification Summary:
+
+### List of Notification Sources
+
+The following list contains notifications and the intended recipients.
+An asterisk by the Purpose field signifies that it's not clear whether the notification
+is needed.
+
+_**Authentication**_
+
+> It needs to be determined whether the Data Steward here is Central or Local
+
+> **IVA**, or **Independent Verification Address**, refers to some physical or digital
+> address, such as a mailing address or phone number, and is used to issue a code for
+> authentication.
+
+| Recipient    | Purpose                                  | Source Exists | Data Req'd |
+|--------------|------------------------------------------|---------------|------------|
+| Data Steward | LS Login doesn’t match what's registered | No            | User ID    |
+| User         | IVA invalidated                          | No            | User ID    |
+| User         | IVA verification code requested          | No            | User ID    |
+| Data Steward | IVA verification code requested          | No            | User ID    |
+| User         | IVA verification code transmitted        | No            | User ID    |
+| Data Steward | IVA verification code submitted by user  | No            | User ID    |
+
+
+_**Data Submission**_
+
+> For the "Research data upload completion" notification, the RDC email must be
+> retrievable from the File ID.
+
+| Recipient     | Purpose                           | Source Exists | Data Req'd |
+|---------------|-----------------------------------|---------------|------------|
+| Central DS    | *Metadata is ready for review     | No            | File ID    |
+| RD Controller | Research data upload completion   | Yes           | File ID    |
+| RD Submitter  | *Approval/rejection of submission | No            | User ID    |
+
+
+
+_**Data Request and Download**_
+
+> If there is a stored entity linking the request to both the dataset and user IDs, then
+> the request would be the only piece of information needed from the event.
+
+| Recipient    | Purpose                          | Source Exists | Data Req'd |
+|--------------|----------------------------------|---------------|------------|
+| DRR          | Request Created (Confirmation)   | No            | Dataset ID, User ID|
+| User (DACR)  | Request Created                  | No            | Dataset ID, User ID|
+| DRR          | Request Allowed                  | No            | Dataset ID, User ID|
+| User (DACR)  | Request Allowed                  | No            | Dataset ID, User ID|
+| DRR          | Request Denied                   | No            | Dataset ID, User ID|
+| User (DACR)  | Request Denied                   | No            | Dataset ID, User ID|
+| DRR          | *Dataset ready for download      | Yes           | Dataset ID, User ID|
```

Comment:
Staging might take some time, I would include this one as needed.
Might make for a better workflow than the connector just idling

Reply:
That was my hunch. At the moment there's no way to tie the file staging event to the user, so that will require investigation. Maybe we could use an header similar to the correlation ID to easily propagate user IDs throughout request flows?